pkg/clientaccess/token_test.go

```diff
@@ -0,0 +1,318 @@
+		assert.EqualError(err, testCase.expected, testCase)
+		assert.Nil(info, testCase)
+
+		info, err = ParseAndValidateTokenForUser(testCase.server, testCase.token, defaultUsername)
+		assert.EqualError(err, testCase.expected, testCase)
+		assert.Nil(info, testCase)
+	}
+}
+
+// TestInvalidCredentials tests that tokens which don't have valid credentials are rejected
+func TestInvalidCredentials(t *testing.T) {
+	assert := assert.New(t)
+
+	server := newTLSServer(t, defaultUsername, defaultPassword)
+	defer server.Close()
+
+	testInfo := &Info{
+		CACerts:  getServerCA(server),
+		BaseURL:  server.URL,
+		Username: "nobody",
+		Password: "invalid",
+		caHash:   hashCA(getServerCA(server)),
+	}
+
+	testCases := []string{
+		testInfo.Password,
+		testInfo.String(),
+	}
+
+	for _, testCase := range testCases {
+		info, err := ParseAndValidateToken(server.URL, testCase)
+		assert.NoError(err, testCase)
+		if assert.NotNil(info) {
+			res, err := Get("/v1-k3s/server-bootstrap", info)
+			assert.Error(err, testCase)
+			assert.Empty(res, testCase)
+		}
+
+		info, err = ParseAndValidateTokenForUser(server.URL, testCase, defaultUsername)
+		assert.NoError(err, testCase)
+		if assert.NotNil(info) {
+			res, err := Get("/v1-k3s/server-bootstrap", info)
+			assert.Error(err, testCase)
+			assert.Empty(res, testCase)
+		}
+	}
+}
+
+// TestConnectionFailures tests that connections are timed out properly
+func TestConnectionFailures(t *testing.T) {
+	testDuration := (defaultClientTimeout * 2) + time.Second
+	assert := assert.New(t)
+	testCases := []struct {
+		server string
+		token  string
+	}{
+		{"https://192.0.2.1:6443", "token"}, // RFC 5735 TEST-NET-1 for use in documentation and example code
+		{"https://localhost:1", "token"},
+	}
+
+	for _, testCase := range testCases {
+		startTime := time.Now()
+		info, err := ParseAndValidateToken(testCase.server, testCase.token)
+		assert.Error(err, testCase)
+		assert.Nil(info, testCase)
+		assert.WithinDuration(time.Now(), startTime, testDuration, testCase)
+
+		startTime = time.Now()
+		info, err = ParseAndValidateTokenForUser(testCase.server, testCase.token, defaultUsername)
+		assert.Error(err, testCase)
+		assert.Nil(info, testCase)
+		assert.WithinDuration(startTime, time.Now(), testDuration, testCase)
+	}
+}
+
+// newTLSServer returns a HTTPS server that mocks the basic functionality required to validate K3s join tokens.
+// Each call to this function will generate new CA and server certificates unique to the returned server.
+func newTLSServer(t *testing.T, username, password string) *httptest.Server {
+	var server *httptest.Server
+	server = httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == "/v1-k3s/server-bootstrap" {
+			if authUsername, authPassword, ok := r.BasicAuth(); ok != true || authPassword != password || authUsername != username {
+				http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
```

Comment:
nit: `StatusUnauthorized`.

Reply:
The real code currently returns a server error in case of bad credentials, so the mock does the same thing.